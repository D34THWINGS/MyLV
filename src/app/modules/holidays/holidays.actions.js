import { fetchWithAuth } from '../auth/auth.actions'

export const HOLIDAYS_FETCH_START = 'HOLIDAYS_FETCH_START'
export const HOLIDAYS_FETCH_SUCCESS = 'HOLIDAYS_FETCH_SUCCESS'
export const HOLIDAYS_FETCH_ERROR = 'HOLIDAYS_FETCH_ERROR'
export const fetchHolidays = () => (dispatch) => {
  dispatch({ type: HOLIDAYS_FETCH_START })

  return dispatch(fetchWithAuth('/api/holidays'))
    .then(holidays => dispatch({ type: HOLIDAYS_FETCH_SUCCESS, payload: holidays }))
    .catch(e => dispatch({ type: HOLIDAYS_FETCH_ERROR, payload: e }))
}

export const PERSONAL_HOLIDAYS_FETCH_START = 'PERSONAL_HOLIDAYS_FETCH_START'
export const PERSONAL_HOLIDAYS_FETCH_SUCCESS = 'PERSONAL_HOLIDAYS_FETCH_SUCCESS'
export const PERSONAL_HOLIDAYS_FETCH_ERROR = 'PERSONAL_HOLIDAYS_FETCH_ERROR'
export const fetchPersonalHolidays = () => (dispatch) => {
  dispatch({ type: PERSONAL_HOLIDAYS_FETCH_START })

  return dispatch(fetchWithAuth('/api/holidays/me'))
    .then(holidays => dispatch({ type: PERSONAL_HOLIDAYS_FETCH_SUCCESS, payload: holidays }))
    .catch(e => dispatch({ type: PERSONAL_HOLIDAYS_FETCH_ERROR, payload: e }))
}

export const postHoliday = holiday => dispatch => dispatch(fetchWithAuth('/api/holidays', {
  method: 'POST',
  body: holiday,
}))

export const putHoliday = ({ id, ...holiday }) => dispatch => dispatch(fetchWithAuth(`/api/holidays/${id}`, {
  method: 'PUT',
  body: holiday,
}))

export const HOLIDAY_DELETE_SUCCESS = 'HOLIDAY_DELETE_SUCCESS'
export const deleteHoliday = id => dispatch =>
  dispatch(fetchWithAuth(`/api/holidays/${id}`, { method: 'DELETE' }))
    .then(() => dispatch({ type: HOLIDAY_DELETE_SUCCESS, payload: { id } }))
