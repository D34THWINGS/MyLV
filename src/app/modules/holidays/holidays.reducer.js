import {
  HOLIDAY_DELETE_SUCCESS,
  HOLIDAY_DETAILS_FETCH_ERROR,
  HOLIDAY_DETAILS_FETCH_START,
  HOLIDAY_DETAILS_FETCH_SUCCESS,
  HOLIDAYS_FETCH_ERROR,
  HOLIDAYS_FETCH_START,
  HOLIDAYS_FETCH_SUCCESS,
  PERSONAL_HOLIDAYS_FETCH_ERROR, PERSONAL_HOLIDAYS_FETCH_START,
  PERSONAL_HOLIDAYS_FETCH_SUCCESS,
} from './holidays.actions'

const initialState = {
  holidaysById: {},
  personalHolidays: [],
  isPersonalLoading: false,
  partnersHolidays: [],
  isPartnersHolidaysLoading: false,
  isUniqueLoading: false,
  limit: 25,
  pageCount: 0,
}

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case PERSONAL_HOLIDAYS_FETCH_START:
      return {
        ...state,
        personalHolidays: [],
        isPersonalLoading: true,
      }
    case PERSONAL_HOLIDAYS_FETCH_SUCCESS:
      return {
        ...state,
        holidaysById: {
          ...state.holidaysById,
          ...payload.results.reduce((acc, holiday) => ({ ...acc, [holiday.id]: holiday }), {}),
        },
        personalHolidays: payload.results.map(holiday => holiday.id),
        isPersonalLoading: false,
      }
    case PERSONAL_HOLIDAYS_FETCH_ERROR:
      return {
        ...state,
        isPersonalLoading: false,
      }
    case HOLIDAY_DELETE_SUCCESS: {
      const { [payload.id]: _, ...holidaysById } = state.holidaysById
      return {
        ...state,
        holidaysById,
        personalHolidays: state.personalHolidays.filter(holidayId => holidayId !== payload.id),
      }
    }
    case HOLIDAYS_FETCH_START:
      return {
        ...state,
        partnersHolidays: [],
        isPartnersHolidaysLoading: true,
        limit: payload.limit || state.limit,
      }
    case HOLIDAYS_FETCH_SUCCESS:
      return {
        ...state,
        holidaysById: {
          ...state.holidaysById,
          ...payload.results.reduce((acc, holiday) => ({ ...acc, [holiday.id]: holiday }), {}),
        },
        partnersHolidays: payload.results.map(holiday => holiday.id),
        isPartnersHolidaysLoading: false,
        pageCount: payload.pageCount,
      }
    case HOLIDAYS_FETCH_ERROR:
      return {
        ...state,
        isPartnersHolidaysLoading: false,
      }
    case HOLIDAY_DETAILS_FETCH_START:
      return {
        ...state,
        isUniqueLoading: true,
      }
    case HOLIDAY_DETAILS_FETCH_SUCCESS:
      return {
        ...state,
        holidaysById: {
          ...state.holidaysById,
          [payload.id]: payload,
        },
        isUniqueLoading: false,
      }
    case HOLIDAY_DETAILS_FETCH_ERROR:
      return {
        ...state,
        isUniqueLoading: false,
      }
    default:
      return state
  }
}
