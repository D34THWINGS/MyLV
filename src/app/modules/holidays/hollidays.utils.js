import moment from 'moment'

import { publicHolidays } from '../../../shared/calendar-constants'

export const formatPeriod = ({
  _id,
  startOnPM,
  endOnPM,
  ...period
}) => ({
  ...period,
  startDate: startOnPM ? moment(period.startDate).hour(11).toISOString() : moment(period.startDate).toISOString(),
  endDate: endOnPM ? moment(period.endDate).hour(23).toISOString() : moment(period.endDate).hour(11).toISOString(),
})

const millisecondsInDay = 1000 * 60 * 60 * 24

export const getPeriodDayCount = (period) => {
  const startDate = moment(period.startDate)
  const endDate = moment(period.endDate)
  let days = Math.round((endDate.diff(startDate) / millisecondsInDay) * 2) / 2
  if (days < 0) {
    return 0
  }
  while (startDate.format('YYYY-MM-DD') !== endDate.format('YYYY-MM-DD')) {
    const month = `0${startDate.month() + 1}`.slice(-2)
    const day = `0${startDate.date()}`.slice(-2)
    if (startDate.day() >= 5 || publicHolidays.has(`${month}-${day}`)) {
      days -= 1
    }
    startDate.add(1, 'd')
  }
  return days
}

export const getDaysForLabel = (periods, key, shouldFormat = true) => periods
  .filter(period => period.label === key)
  .map(shouldFormat ? formatPeriod : period => period)
  .reduce((sum, period) => sum + getPeriodDayCount(period), 0)
