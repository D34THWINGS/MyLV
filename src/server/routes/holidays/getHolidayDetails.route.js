const Boom = require('boom')
const { cracra } = require('config')

const lvConnect = require('../../helpers/lvconnect.helper')

module.exports = {
  method: 'GET',
  path: '/api/holidays/{id}',
  async handler(req, res) {
    const holidayRequest = await req.server.app.models.Holiday.findById(req.params.id)

    const canViewOthersRequests = cracra.partnersRoles.some(role => req.auth.credentials.roles.indexOf(role) >= 0)
    if (holidayRequest.user.toString() !== req.auth.credentials.id && !canViewOthersRequests) {
      return res(Boom.forbidden('Insufficient rights.'))
    }

    const response = await lvConnect.api(`/users/${holidayRequest.user}`)
    const partner = await response.json()

    if (!holidayRequest) {
      res(Boom.notFound('Holiday request not found.'))
    } else {
      res.mongodb({
        ...holidayRequest.toJSON(),
        user: partner,
      })
    }
  },
}
