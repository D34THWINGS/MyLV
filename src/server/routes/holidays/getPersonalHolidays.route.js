module.exports = {
  method: 'GET',
  path: '/api/holidays/me',
  async handler(req, res) {
    const holidays = await req.server.app.models.Holiday
      .where('user', req.auth.credentials.id)
      .sort('-date')
      .find()

    res.mongodb({ results: holidays })
  },
}
