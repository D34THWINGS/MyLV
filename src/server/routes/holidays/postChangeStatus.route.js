const Joi = require('joi')
const config = require('config')

const hasRole = require('../../helpers/hasRole.pre')

const statusMapping = {
  approved: 'acceptée',
  rejected: 'refusée',
}

module.exports = {
  method: 'POST',
  path: '/api/holidays/{id}/changeStatus',
  config: {
    validate: {
      payload: {
        status: Joi.string().valid(['approved', 'rejected']).required(),
      },
    },
    pre: [hasRole(config.cracra.partnersRoles)],
  },
  async handler(req, res) {
    const { status } = req.payload
    const holidayRequest = await req.server.app.models.Holiday.findOneAndUpdate(
      { _id: req.params.id },
      { $set: { status } },
    )

    req.server.methods.sendPushNotification(holidayRequest.user, JSON.stringify({
      message: `Demande de congé "${holidayRequest.title}" ${statusMapping[status]}`,
      url: `/holidays/${holidayRequest._id}`,
    }))

    res.mongodb({ ...holidayRequest.toJSON(), status })
  },
}
