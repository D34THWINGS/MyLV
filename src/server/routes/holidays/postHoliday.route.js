const Joi = require('joi')
const moment = require('moment')
const { holidayLabels } = require('../../../shared/calendar.constants')

module.exports = {
  method: 'POST',
  path: '/api/holidays',
  config: {
    validate: {
      payload: {
        title: Joi.string().max(255).required(),
        comment: Joi.string().max(255),
        periods: Joi.array().items(Joi.object({
          label: Joi.string().valid(Array.from(holidayLabels.keys())).required(),
          startDate: Joi.date().required(),
          endDate: Joi.date().required(),
        })).required(),
      },
    },
  },
  async handler(req, res) {
    const holiday = await req.server.app.models.Holiday.create(Object.assign(req.payload, {
      user: req.auth.credentials.id,
      date: moment().toDate(),
    }))

    req.server.plugins.mailjet.sendHolidaysRequest(req.auth.credentials, holiday)

    res.mongodb(holiday)
  },
}
