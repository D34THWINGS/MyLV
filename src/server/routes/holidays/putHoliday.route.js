const Joi = require('joi')
const { holidayLabels } = require('../../../shared/calendar.constants')

module.exports = {
  method: 'PUT',
  path: '/api/holidays/{id}',
  config: {
    validate: {
      payload: {
        title: Joi.string().max(255).required(),
        comment: Joi.string().max(255),
        periods: Joi.array().items(Joi.object({
          label: Joi.string().valid(Array.from(holidayLabels.keys())).required(),
          startDate: Joi.date().required(),
          endDate: Joi.date().required(),
        })).required(),
      },
    },
  },
  async handler(req, res) {
    const holiday = await req.server.app.models.Holiday.findOneAndUpdate({ _id: req.params.id }, { $set: req.payload })

    res.mongodb(holiday)
  },
}
