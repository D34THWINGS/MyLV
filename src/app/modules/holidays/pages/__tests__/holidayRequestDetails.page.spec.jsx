import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import { HolidayRequestDetails } from '../holidayRequestDetails.page'
import { getPeriodDayCount } from '../../../../../shared/holidays.utils'

jest.unmock('../holidayRequestDetails.page')

describe('HolidayRequestPage', () => {
  let props
  beforeEach(() => {
    props = {
      fetchHolidayRequestDetails: jest.fn(),
      changeHolidayRequestStatus: jest.fn(),
      holidayRequest: {
        user: 'foo',
        title: 'hello',
        comment: 'world',
        periods: [{
          _id: 'qux',
          label: 'paidHolidays',
          startDate: '2012-01-01',
          endDate: '2012-01-02',
        }],
      },
      user: { id: 'foo' },
      isLoading: false,
      match: {
        params: {
          id: 'bar',
        },
      },
      partner: {
        firstName: 'james',
        lastName: 'bond',
      },
      classes: {},
    }

    getPeriodDayCount.mockImplementation(() => 42)
  })

  const getWrapper = () => shallow(<HolidayRequestDetails {...props} />)

  it('should render properly', () => {
    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render in loading state', () => {
    // Given
    props.isLoading = true

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render not found', () => {
    // Given
    props.holidayRequest = null

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render other\'s holiday request', () => {
    // Given
    props.holidayRequest.user = 'bar'

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
