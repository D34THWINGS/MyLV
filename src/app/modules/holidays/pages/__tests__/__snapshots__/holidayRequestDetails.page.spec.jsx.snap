// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HolidayRequestPage should render in loading state 1`] = `<WithStyles(LoadingPage) />`;

exports[`HolidayRequestPage should render not found 1`] = `<WithStyles(NotFound) />`;

exports[`HolidayRequestPage should render other's holiday request 1`] = `
<Card
  raised={false}
>
  <WithStyles(CardContent)>
    <div>
      <WithStyles(Typography)
        component="h2"
        gutterBottom={true}
        variant="headline"
      >
        Demande de congé de james bond
      </WithStyles(Typography)>
      <HolidayRequestStatusIcon
        status="pending"
      />
    </div>
    <WithStyles(Typography)
      component="p"
      gutterBottom={true}
      variant="body1"
    >
      world
    </WithStyles(Typography)>
    <WithStyles(List)>
      <WithStyles(ListSubheader)>
        Période(s) demandée(s)
      </WithStyles(ListSubheader)>
      <WithStyles(ListItem)
        key="qux"
      >
        <WithStyles(Avatar)>
          <pure(BeachAccess) />
        </WithStyles(Avatar)>
        <WithStyles(ListItemText)
          primary="Du 01/01/2012 au 02/01/2012 (42 jours)"
          secondary="Congés payés"
        />
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(CardContent)>
  <Restricted
    roles={
      Array [
        "hr",
        "board",
      ]
    }
    user={
      Object {
        "id": "foo",
      }
    }
  >
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        onClick={[Function]}
        size="small"
      >
        Approuver
      </WithStyles(Button)>
      <WithStyles(Button)
        color="default"
        onClick={[Function]}
        size="small"
      >
        Refuser
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </Restricted>
</Card>
`;

exports[`HolidayRequestPage should render properly 1`] = `
<Card
  raised={false}
>
  <WithStyles(CardContent)>
    <div>
      <WithStyles(Typography)
        component="h2"
        gutterBottom={true}
        variant="headline"
      >
        hello
      </WithStyles(Typography)>
      <HolidayRequestStatusIcon
        status="pending"
      />
    </div>
    <WithStyles(Typography)
      component="p"
      gutterBottom={true}
      variant="body1"
    >
      world
    </WithStyles(Typography)>
    <WithStyles(List)>
      <WithStyles(ListSubheader)>
        Période(s) demandée(s)
      </WithStyles(ListSubheader)>
      <WithStyles(ListItem)
        key="qux"
      >
        <WithStyles(Avatar)>
          <pure(BeachAccess) />
        </WithStyles(Avatar)>
        <WithStyles(ListItemText)
          primary="Du 01/01/2012 au 02/01/2012 (42 jours)"
          secondary="Congés payés"
        />
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(CardContent)>
  <Restricted
    roles={
      Array [
        "hr",
        "board",
      ]
    }
    user={
      Object {
        "id": "foo",
      }
    }
  >
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        onClick={[Function]}
        size="small"
      >
        Approuver
      </WithStyles(Button)>
      <WithStyles(Button)
        color="default"
        onClick={[Function]}
        size="small"
      >
        Refuser
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </Restricted>
</Card>
`;
